### ISSUE TITLE
question: liburing access patterns

### ISSUE THREAD

**Issue by romange** [axboe/liburing#103]
1. I am looking for  a best-practice to "interrupt" the `io_uring_enter` call from a different thread. I am using eventfd which works ok. it seems that IORING_OP_NOP could work as well if `submit` path was thread-safe but it seems not. Is it non-thread safe in kernel as well? Can you think of anything  better than eventfd for waking up io_uring_enter from a different thread? btw, I am getting a hard time to understand why IORING_REGISTER_EVENTFD exists. Does it help me it if I manage my I/O loop fully via io_uring (without epoll_wait)? 
2. `io_uring_peek_batch_cqe` returns `count` *pointers* to entries in `ring->cq.cqes` array. The entries are 16 bytes, pointers are 8. Why not just copy the entries by value? Why do we need this additional level of indirection? That would allow a user to advance completion queue by `count` right after `io_uring_peek_batch_cqe` returns and before it processes cqe callbacks potentially freeing up the space in CQ for kernel  in parallel to user code processing the callbacks.

**Comment by axboe** [comment]
1. eventfd should work fine, another option is to use timeouts
2. you must process cqes before advancing the cq ring, or you could risk have new events come in an overwrite the ones you haven't processed yet. if you copy, then you need an extra copy (both in terms of storage and memory copy overhead). you could roll that on your side if you wanted to, though.

**Comment by axboe** [comment]
To expand on #1, the kernel side is fully thread side, the liburing side is not in terms of the ```struct io_uring``` cq and sq side management.

**Comment by romange** [comment]
Thanks! My suggestion regarding (2) was to simplify the interface since the extra copy is only 8 bytes per entry and it seems negligible. Anyway, that's what I did https://github.com/romange/gaia/blob/IoUring/util/uring/proactor.cc#L55

**Comment by jorangreef** [comment]
@axboe, @romange 's suggestion regarding copying 16 byte CQEs instead of copying 8 byte pointers would also mean that `io_uring_cqe_seen()` is no longer needed, making the liburing interface simpler and safer. It might even be faster... if you could copy a batch of CQEs, you would also gain the performance upside of amortizing the atomic head update across the batch.

For a batch of 256 CQEs, or a 4096 page, it comes down to a copy of 2048 additional bytes with a single `io_uring_smp_store_release()` vs 2048 less bytes copied with 256 calls to `io_uring_smp_store_release()`.

I don't know which is worse.

**Comment by axboe** [comment]
I've got some pretty beefy test hardware, so I'd definitely be able to test this if someone hacks it up. 

On difference in cost, depends a lot on the arch. On x86, the load acquire and store release are not expensive.

**Comment by isilence** [comment]
Well, you're right, but I don't like `io_uring_peek_batch_cqe()` at all.
CQEs are already laying mostly contiguously in memory with an exception
of wraping around. You may iterating over them directly (see
`io_uring_for_each_cqe()`) or do maximum 2 memcpy()s to copy them
to a local array of CQEs.

p.s. io_uring_for_each_cqe() does unnecessary load_acquire()s, but
that's doing nothing for x86 and you may easily optimise it.

**Comment by romange** [comment]
I perform the copy to local array and then call `io_uring_cq_advance`  *before* running the completion callbacks. The motivation is that I want to remove pending CQEs as quickly as possible to allow kernel to continue processing I/O events  while my server code does its own thing. 

```
inline unsigned CQReadyCount(const io_uring& ring) {
  return io_uring_smp_load_acquire(ring.cq.ktail) - *ring.cq.khead;
}

unsigned IoRingPeek(const io_uring& ring, io_uring_cqe* cqes, unsigned count) {
  unsigned ready = CQReadyCount(ring);
  if (!ready)
    return 0;

  count = count > ready ? ready : count;
  unsigned head = *ring.cq.khead;
  unsigned mask = *ring.cq.kring_mask;
  unsigned last = head + count;
  for (int i = 0; head != last; head++, i++) {
    cqes[i] = ring.cq.cqes[head & mask];
  }
  return count;
}
```

**Comment by isilence** [comment]
IMHO, this one should be in liburing.

And that may be faster with memcpy().
```C
unsigned head, mask, last;
...
count = min(ready, count);
last = min(head + count, (head + mask) & mask);
memcpy(local_cqes, &ring.cq.cqes[head & mask], last - head);
```

**Comment by isilence** [comment]
Or this one

```C
offset = head & mask;
cq_size = mask + 1;
count = min(ready, count, cq_size - offset);
memcpy(local_cqes, &ring.cq.cqes[head & mask], count);
```

**Comment by romange** [comment]
yes, it's faster and more concise. 
as you said, it might need 2 memcpy's in case of wrapping.

**Comment by isilence** [comment]
Right, that one should work if callers do retries for non-zero return.
You're probably already have a loop around, am I wrong?

**Comment by romange** [comment]
Yes, I have a loop around :+1:  Good point.
